backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.backup.s3;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.camunda.zeebe.backup.api.Backup;
+import io.camunda.zeebe.backup.api.BackupIdentifier;
+import io.camunda.zeebe.backup.api.BackupStatus;
+import io.camunda.zeebe.backup.api.BackupStore;
+import java.util.concurrent.CompletableFuture;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+
+/**
+ * {@link BackupStore} for S3. Stores all backups in a given bucket.
+ *
+ * <p>All created object keys are prefixed by the {@link BackupIdentifier}, with the following
+ * scheme: {@code checkpointId/partitionId/nodeId}
+ *
+ * <p>Each backup contains:
+ *
+ * <ol>
+ *   <li>A 'metadata' object, containing {@link Metadata} serialized as JSON, for example
+ *       <pre>checkpointId/partitionId/nodeId/metadata</pre>
+ *   <li>A 'status' object, containing the {@link BackupStatus}, for example
+ *       <pre>checkpointId/partitionId/nodeId/status</pre>
+ *   <li>Objects for snapshot files, additionally prefixed with 'snapshot', for example
+ *       <pre>checkpointId/partitionId/nodeId/snapshots/snapshot-file-1</pre>
+ *   <li>Objects for segment files, additionally prefixed with 'segments', for example
+ *       <pre>checkpointId/partitionId/nodeId/segments/segment-file-1</pre>
+ * </ol>
+ */
+public final class S3BackupStore implements BackupStore {
+
+  static final ObjectMapper MAPPER = new ObjectMapper();
+
+  private final S3Config config;
+  private final S3AsyncClient client;
+
+  public S3BackupStore(final S3Config config, final S3AsyncClient client) {
+    this.config = config;
+    this.client = client;
+  }
+
+  @Override
+  public CompletableFuture<Void> save(final Backup backup) {
+    final var metadata = saveMetadata(backup);
+
+    return CompletableFuture.allOf(metadata);
+  }
```

Comment:
:question:  Should this run on a separate thread to not block the caller? Or is it enough that client is async which uses its own executor?

The caller is BackupManager which is running on its own actor. I think it is better to not block it so that it can do other operations such as  responding to status query while taking a backup.

Reply:
This is something I haven't thought about much until now. The S3 async client uses it's own thread pool for IO but eventually the backup store will also have to wait for the upload to complete update the status. This is where we will have to be careful to use the same thread pool and not the callers thread. 

I think for now we can keep this as is, returning a CompletableFuture that is completed by the s3 client when uploading is finished is exactly what we want to do here. BackupManager can decide if it wants to wait for completion or not.